core/src/test/scala/shapeless/constraints.scala

```diff
@@ -0,0 +1,486 @@
+/*
+ * Copyright (c) 2011-2020 Miles Sabin 
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package shapeless
+
+import org.junit.Test
+import org.junit.Assert._
+
+import shapeless.test.illTyped
+
+trait Fruit
+case object Apple extends Fruit
+case object Pear extends Fruit
+
+class UnaryTCConstraintTests{
+  import UnaryTCConstraint._
+
+  def acceptOption[T](l : T)(implicit ev: UnaryTCConstraint[T, Option]) = true
+  def acceptId[T](l : T)(implicit ev: UnaryTCConstraint[T, Id]) = true
+  def acceptConst[T](l : T)(implicit ev: UnaryTCConstraint[T, Const[String]#λ]) = true
+  def acceptTypeConstructor[F[_], T](l : T)(implicit ev: UnaryTCConstraint[T, F]) = true
+
+  @Test
+  def testHList: Unit ={
+    val l1 = Option(23) :: Option(true) :: Option("foo") :: HNil 
+    val l2 = Option(23) :: true :: Option("foo") :: HNil
+
+    acceptOption(l1)  // Compiles
+    acceptOption(HNil: HNil)
+
+    illTyped("""
+    acceptOption(l2)
+    """)
+
+    val l3 = 23 :: true :: "foo" :: HNil 
+
+    acceptId(l3)  // Compiles
+    acceptId(HNil: HNil)
+
+    val l4 = "foo" :: "bar" :: "baz" :: HNil
+    val l5 = "foo" :: true :: "baz" :: HNil
+
+    acceptConst(l4)  // Compiles
+    acceptConst(HNil: HNil)
+    illTyped("""
+    acceptConst(l5)
+    """)
+
+    acceptTypeConstructor(l1)  // Compiles - F = Option
+    acceptTypeConstructor(l2)  // Compiles - F = Id
+    acceptTypeConstructor(l3)  // Compiles - F = Id
+    acceptTypeConstructor(l4)  // Compiles - F = Const[String]
+    acceptTypeConstructor(l5)  // Compiles - F = Id
+    acceptTypeConstructor(HNil: HNil)
+  }
+
+  @Test
+  def testTuple: Unit ={
+    val l1 = (Option(23), Option(true), Option("foo"))
+    val l2 = (Option(23), true, Option("foo"))
+
+    acceptOption(l1)  // Compiles
+
+    illTyped("""
+    acceptOption(l2)
+    """)
+
+    val l3 = (23, true, "foo")
+
+    acceptId(l3)  // Compiles
+
+    val l4 = ("foo", "bar", "baz")
+    val l5 = ("foo", true, "baz")
+
+    //TODO: ok, not sure why this isn't working
+    //acceptConst(l4)  // Doesn't compile, D'Oh!
```

Comment:
@wheaties to make this compile you would need to add a special instance for `Const` and `Generic`.
The reason is type inference on Scala < 2.13 can't unify this type.
Something like this:
```scala
 implicit def genericConstTC[A, R, C](
    implicit gen: Generic.Aux[A, R], utc: UnaryTCConstraint[R, Const[C]#λ]
  ): UnaryTCConstraint[A, Const[C]#λ] = instance[A, Const[C]#λ]
```

Reply:
Brilliant. Probably the one thing I didn't try. Let me see what I can do with the rest of your comments.